src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.hpp

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include <csignal>
+#include <iostream>
+#include <vector>
+
+#include "CoreTypes.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+
+    // TODO: Remove this! Use standard ones.
+    static void FailNow(const char* message) { std::cout << message; throw message; }
+    static void FailIf(bool condition, const char* message) {
+        if (condition) {
+            FailNow(message);
+        }
+    }
+
+
+    // TODO: Decide on the interface and move this to appropriate place.
+    struct IQubitManager
+    {
+        virtual ~IQubitManager() = default;
+        virtual Qubit Allocate() = 0;
+        virtual void Release(Qubit qubit) = 0;
+    };
+
+    // The end of free lists are marked with NoneMarker value. It is used like null for pointers.
+    // This value is non-negative just like other values in the free lists.
+    constexpr int NoneMarker = INT_MAX;
```

Comment:
Nit: In a number of places the _signed_ integer types are used (for an _unsigned_ concept like array index or number of elements) and then checked against `< 0`. Consider the _unsigned_ integer types (and remove the checks `< 0`).

FYI: For my own code I consider _unsigned_ integer types by default.

Reply:
Long time ago I was also advocating unsigned types for unsigned concepts. I changed my mind after seeing how convenient signed types are for these concepts in C#. I will make a separate comment about this later.